mm2src/coins/qrc20/swap.rs

```diff
@@ -113,46 +113,51 @@ impl Qrc20Coin {
         secret_hash: Vec<u8>,
         amount: BigDecimal,
         expected_swap_contract_address: H160,
-    ) -> Result<(), String> {
+    ) -> Result<(), MmError<ValidatePaymentError>> {
         let expected_swap_id = qrc20_swap_id(time_lock, &secret_hash);
-        let status = try_s!(
-            self.payment_status(&expected_swap_contract_address, expected_swap_id.clone())
-                .await
-        );
+        let status = self
+            .payment_status(&expected_swap_contract_address, expected_swap_id.clone())
+            .await
+            .map_to_mm(ValidatePaymentError::TransportError)?;
         if status != eth::PAYMENT_STATE_SENT.into() {
-            return ERR!("Payment state is not PAYMENT_STATE_SENT, got {}", status);
+            return MmError::err(ValidatePaymentError::UnexpectedPaymentState(format!(
+                "Payment state is not PAYMENT_STATE_SENT, got {}",
+                status
+            )));
         }
 
         let expected_call_bytes = {
-            let expected_value = try_s!(wei_from_big_decimal(&amount, self.utxo.decimals));
-            let my_address = try_s!(self.utxo.derivation_method.iguana_or_err()).clone();
-            let expected_receiver = try_s!(qtum::contract_addr_from_utxo_addr(my_address));
-            try_s!(self.erc20_payment_call_bytes(
+            let expected_value = wei_from_big_decimal(&amount, self.utxo.decimals)?;
+            let my_address = self.utxo.derivation_method.iguana_or_err()?.clone();
+            let expected_receiver = qtum::contract_addr_from_utxo_addr(my_address)?;
+            self.erc20_payment_call_bytes(
                 expected_swap_id,
                 expected_value,
                 time_lock,
                 &secret_hash,
-                expected_receiver
-            ))
+                expected_receiver,
+            )?
         };
-
-        let erc20_payment = try_s!(self.erc20_payment_details_from_tx(&payment_tx).await);
+        let erc20_payment = self
+            .erc20_payment_details_from_tx(&payment_tx)
+            .await
+            .map_to_mm(ValidatePaymentError::UnexpectedErc20PaymentData)?;
```

Comment:
I think you can get rid of `UnexpectedErc20PaymentData` and use `InvalidPaymentTxData`

Reply:
done